### PR TITLE
Expand support of types handled in EXPLAIN (TYPE IO)

### PR DESCRIPTION
**Pull request by JamesRTaylor**
Currently the EXPLAIN (TYPE IO) command has very limited support for column types. This PR expands that to include the primitive column types.

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/PlanPrinter.java [hunk 1]
```diff
@@ -282,7 +277,7 @@ private static String formatFragment(FunctionRegistry functionRegistry, Session
                 .map(argument -> {
                     if (argument.isConstant()) {
                         NullableValue constant = argument.getConstant();
-                        String printableValue = castToVarchar(constant.getType(), constant.getValue(), functionRegistry, session);
+                        String printableValue = PlanPrinterUtil.castToVarchar(constant.getType(), constant.getValue(), functionRegistry, session);
```
Comment (ebyhr) [inline presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/PlanPrinter.java]: You can import as
```java
import static io.prestosql.sql.planner.planPrinter.PlanPrinterUtil.castToVarchar;
```

File: presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/PlanPrinterUtil.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import io.prestosql.Session;
+import io.prestosql.metadata.FunctionRegistry;
+import io.prestosql.metadata.OperatorNotFoundException;
+import io.prestosql.metadata.Signature;
+import io.prestosql.spi.type.Type;
+import io.prestosql.sql.InterpretedFunctionInvoker;
+
+import static io.prestosql.spi.type.VarcharType.VARCHAR;
+
+public class PlanPrinterUtil
```
Comment (ebyhr) [inline presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/PlanPrinterUtil.java]: It would be better to use `final` for util class.

File: presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/PlanPrinterUtil.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+{
+    private PlanPrinterUtil()
+    {
+    }
```
Comment (ebyhr) [inline presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/PlanPrinterUtil.java]: nit: Most util classes in presto use below syntax.
```java
private PlanPrinterUtil() {}
```

**Comment by ebyhr** [comment]
Left few minor comments. Thank you for creating this PR!

**Comment by JamesRTaylor** [comment]
Thanks for the feedback, @ebyhr. I've pushed commits to address them.

**Comment by martint** [comment]
@cla-bot check

File: presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/IoPlanPrinter.java [hunk 4]
```diff
@@ -61,11 +66,21 @@
...
 import static java.util.Objects.requireNonNull;
 
 public class IoPlanPrinter
 {
+    /**
+     * List of supported data types for column constraints appearing in
+     * EXPLAIN (TYPE IO). Other types would cause a NOT_SUPPORTED
+     * Presto exception to be thrown.
+     */
+    private static final Set<Class> WHITELISTED_TYPES_FOR_OUTPUT = ImmutableSet.of(
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/IoPlanPrinter.java]: We could support any type that has a proper CAST to VARCHAR instead of an explicit whitelist.

Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/IoPlanPrinter.java]: So, the implementation of getVarcharValue() below should check if there's a CAST (like in PlanPrinter.castToVarchar) and invoke it or throw the "Unsupported data type in EXPLAIN (TYPE IO)" exception.

Comment (JamesRTaylor) [reply presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/IoPlanPrinter.java]: Thanks for the feedback, @martint. I've pushed a new commit that removed the whitelist and instead relies on whether the CAST to VARCHAR succeeds or fails.

File: presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/IoPlanPrinter.java [hunk 5]
```diff
@@ -579,12 +559,14 @@ private FormattedMarker formatMarker(Marker marker)
...
         {
-            for (Class<Type> whitelistedType : WHITELISTED_TYPES_FOR_OUTPUT) {
-                if (whitelistedType.getClass().isInstance(type.getClass())) {
-                    return castToVarchar(type, value, metadata.getFunctionRegistry(), session);
-                }
+            try {
+                return throwOrCastToVarchar(type, value, metadata.getFunctionRegistry(), session);
+            }
+            catch (OperatorNotFoundException e) {
+                throw new PrestoException(
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/IoPlanPrinter.java]: Unnecessary formatting change. These can all go on the same line.

File: presto-tests/src/test/java/io/prestosql/sql/planner/planPrinter/TestPlanPrinterUtil.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.Optional;
+
+import static io.prestosql.block.BlockAssertions.createLongsBlock;
+import static io.prestosql.block.BlockAssertions.createStringsBlock;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.VarcharType.VARCHAR;
+import static io.prestosql.util.StructuralTestUtil.mapType;
+
+public class TestPlanPrinterUtil
+        extends AbstractTestQueryFramework
```
Comment (martint) [inline presto-tests/src/test/java/io/prestosql/sql/planner/planPrinter/TestPlanPrinterUtil.java]: This doesn't need to extend AbstractTestQueryFramework. That's intended for writing tests that will execute every query against a custom query runner (e.g., to test different connectors or configurations of the query runner).

Comment (martint) [reply presto-tests/src/test/java/io/prestosql/sql/planner/planPrinter/TestPlanPrinterUtil.java]: Also, I think these tests are a little brittle (one of them relies on an implementation detail of the map type) and overkill, as they are mostly testing whether cast to varchar works. I'd just remove them altogether.

File: presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/PlanPrinterUtil.java [hunk 7]
```diff
@@ -29,17 +29,23 @@ private PlanPrinterUtil() {}
...
-            Slice coerced = (Slice) new InterpretedFunctionInvoker(functionRegistry).invoke(coercion, session.toConnectorSession(), value);
-            return coerced.toStringUtf8();
+            return throwOrCastToVarchar(type, value, functionRegistry, session);
         }
         catch (OperatorNotFoundException e) {
             return "<UNREPRESENTABLE VALUE>";
         }
     }
+
+    static String throwOrCastToVarchar(Type type, Object value, FunctionRegistry functionRegistry, Session session)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/planPrinter/PlanPrinterUtil.java]: How about we name this "castToVarcharOrFail" ?

File: presto-tests/src/test/java/io/prestosql/sql/planner/planPrinter/TestPlanPrinterUtil.java (hunk 6 above)
Comment (JamesRTaylor) [reply presto-tests/src/test/java/io/prestosql/sql/planner/planPrinter/TestPlanPrinterUtil.java]: I can remove these tests, no problem, but the reason I added these in the first place was to make sure that the castToVarchar fails for non primitive types. The function requires a Session and FunctionRegistry and I wasn't sure how to get those without using the TestLocalQueries::createLocalQueryRunner. I'm sure there's a better way, though. WDYT?

Comment (martint) [reply presto-tests/src/test/java/io/prestosql/sql/planner/planPrinter/TestPlanPrinterUtil.java]: The issue is that it should fail for any type that doesn't have a cast to varchar. Some complex types may implement a cast. The right way to test this would be to create a synthetic type just for this purpose, but again, I think that's overkill.

Comment (JamesRTaylor) [reply presto-tests/src/test/java/io/prestosql/sql/planner/planPrinter/TestPlanPrinterUtil.java]: Ok, I'll just remove this test then.

**Comment by martint** [comment]
@JamesRTaylor, can you rebase and fix the conflicts so that Travis can run?

**Comment by martint** [comment]
I made a couple of minor adjustments, rebased and merged it. Thanks for your contribution!
